NEW Handle WithinRangeFilter in searchable_fields and SearchContext

GuySartorelli opened the pull request:
> [!IMPORTANT]
> This PR has multiple commits. Do not squash - or if you do, use the second commit's message

This PR enables using `WithinRangeFilter` to filter by numeric, date, and time fields in `searchable_fields`. It can also be used with other field types (see the varchar unit test scenario) if some additional config is provided.

This PR is required so that the current `SiteTree` filter on `LastEdited` can be performed using a `SearchContext` instead of using the current custom weird filter it uses.

Note that this PR does _not_ include any updates to the css e.g. to make date fields look nicer - that will come in https://github.com/silverstripe/silverstripe-admin/pull/1897. For now we're adding just the functionality itself.

See the docs or unit tests for configuration that you can use to test locally.

## Issue
- https://github.com/silverstripe/silverstripe-cms/issues/2949

GuySartorelli committed:
API Provide minimum and maximum comparison values for numeric DBFields

This will be used with handling WithinRangeFilter in searchable fields
so that if only one of the two ends of a range is filled in, the other
can be automatically populated.

GuySartorelli committed:
NEW Handle WithinRangeFilter in searchable_fields and SearchContext

This is especially helpful with date ranges, e.g. all records edited
within a certain range, but can be used out of the box with basically
any numeric, date, or time-based fields.

GuySartorelli commented on src/Core/Validation/FieldValidation/BigIntFieldValidator.php:
```diff
@@ -4,28 +4,14 @@
  */
 class BigIntFieldValidator extends IntFieldValidator
 {
-    /**
-     * The minimum value for a signed 64-bit integer.
-     * Defined as string instead of int otherwise will end up as a float
-     * on 64-bit systems
-     *
-     * When this is cast to an int in IntFieldValidator::__construct()
-     * it will be properly cast to an int
-     */
-    protected const MIN_INT = '-9223372036854775808';
-
-    /**
-     * The maximum value for a signed 64-bit integer.
-     */
-    protected const MAX_INT = '9223372036854775807';
```
These are moved to `DBBigInt` so they can be used with the new `getMinValue()` and `getMaxValue()`.

GuySartorelli commented on src/Core/Validation/FieldValidation/IntFieldValidator.php:
```diff
@@ -3,35 +3,24 @@
  */
 class IntFieldValidator extends NumericNonStringFieldValidator
 {
-    /**
-     * The minimum value for a signed 32-bit integer.
-     * Defined as string instead of int because be cast to a float
-     * on 32-bit systems if defined as an int
-     */
-    protected const MIN_INT = '-2147483648';
-
-    /**
-     * The maximum value for a signed 32-bit integer.
-     */
-    protected const MAX_INT = '2147483647';
```
These are moved to `DBInt` so they can be used with the new `getMinValue()` and `getMaxValue()`.

GuySartorelli commented on src/ORM/DataObject.php:
```diff
@@ -2438,40 +2440,7 @@ public function scaffoldSearchFields($_params = null)
-                    ));
-                }
-            }
+            $field = $this->scaffoldSingleSearchField($fieldName, $spec, $params);
```
This was a lot of code to parse through just to see that it's all for this singular purpose. Made sense to abstract into its own method here to encapsulate this functionality, reducing cognitive load for anyone working in this area in the future.

GuySartorelli commented on src/ORM/DataObject.php:
```diff
@@ -3896,28 +3925,40 @@ public function searchableFields()
         $rewrite = [];
         foreach ($fields as $name => $specOrName) {
             $identifier = (is_int($name)) ? $specOrName : $name;
+            $relObject = $this->relObject($identifier);
 
             if (is_int($name)) {
                 // Format: array('MyFieldName')
                 $rewrite[$identifier] = [];
             } elseif (is_array($specOrName) && (isset($specOrName['match_any']))) {
                 $rewrite[$identifier] = $fields[$identifier];
                 $rewrite[$identifier]['match_any'] = $specOrName['match_any'];
-            } elseif (is_array($specOrName) && ($relObject = $this->relObject($identifier))) {
+            } elseif (is_array($specOrName)) {
```
The condition would cause the following problem:

You have config like this:
```php
[
    'MyRelation' => [
        'filter' => 'ExactMatch',
    ],
]
```

Because this is an array, but the relObject returns null, this condition isn't met. So we go to the `else` clause. This turns the config into this:

```php
[
    'MyRelation' => [
        'filter' => [
            'filter' => 'ExactMatch',
        ],
    ],
]
```

GuySartorelli commented on src/ORM/DataObject.php:
```diff
@@ -3896,28 +3925,40 @@ public function searchableFields()
                 $rewrite[$identifier] = array_merge(
-                    ['filter' => $relObject->config()->get('default_search_filter_class')],
+                    [
+                        'filter' => $relObject?->config()->get('default_search_filter_class'),
+                        'dataType' => $relObject ? get_class($relObject) : null,
```
If there's no relObject, these will just default to `null`. If the `$specOrName` array doesn't fill them in, they'll get filled in further down the line.

GuySartorelli commented on src/ORM/FieldType/DBBigInt.php:
```diff
@@ -14,6 +14,21 @@
  */
 class DBBigInt extends DBInt
 {
+    /**
+     * The minimum value for a signed 64-bit integer.
+     * Defined as string instead of int otherwise will end up as a float
+     * on 64-bit systems
+     *
+     * When this is cast to an int in IntFieldValidator::__construct()
+     * it will be properly cast to an int
+     */
+    protected const MIN_INT = '-9223372036854775808';
+
+    /**
+     * The maximum value for a signed 64-bit integer.
+     */
+    protected const MAX_INT = '9223372036854775807';
```
These override consts of the same name from `DBInt` and are used in the new `getMinValue()` and `getMaxValue()` methods.

GuySartorelli commented on src/ORM/FieldType/DBDate.php:
```diff
@@ -619,4 +619,14 @@ protected function explodeDateString($value)
         return $parts;
     }
+
+    public static function getMinValue(): string
+    {
+        return '0000-00-00';
+    }
+
+    public static function getMaxValue(): string
+    {
+        return '9999-12-31';
+    }
```
These new methods are being added to all date, time, and numeric field types where we know exactly what the minimum and maximum values are. These are intended to be used for comparisons, as noted in the PHPDoc in `DBField`.

GuySartorelli commented on src/ORM/FieldType/DBField.php:
```diff
@@ -571,4 +571,20 @@ public function scalarValueOnly(): bool
         return true;
     }
+
+    /**
+     * @return mixed The minimum value for comparisons with this field - or null if that's not determinable.
+     */
+    public static function getMinValue(): mixed
+    {
+        return null;
+    }
+
+    /**
+     * @return mixed The maximum value for comparisons with this field - or null if that's not determinable.
+     */
+    public static function getMaxValue(): mixed
+    {
+        return null;
+    }
```
Methods with these names already exists on `DBForeignKey` and `DBPercentage`. Adding these method signatures gives us a clear and consistent API (e.g. `DBYear` had `getMinYear()` and `getMaxYear()` instead - now updated to these consistent names) and an easy way to compare the current value against a minimum or maximum value.

See their usage in `DataObject::searchableFields()` setting the default `rangeFromDefault` and `rangeToDefault` values for the reason I'm doing it in this PR. That allows a user to populate only the from OR to field, and get sensible filtered results.

GuySartorelli commented on src/ORM/DataObject.php:
```diff
@@ -2483,6 +2452,24 @@ public function scaffoldSearchFields($_params = null)
+                $originalTitle = $field->Title();
+                $originalName = $field->getName();
+
+                $fieldFrom->setName($originalName . '_SearchFrom');
```
The `_Search` part of this suffix is to help prevent clashes, in case someone has something like a `Price` and a `PriceFrom` DB field - this won't clash with it if they add a range filter for `Price`.

GuySartorelli commented on tests/php/ORM/Search/BasicSearchContextTest.php:
```diff
@@ -339,4 +341,24 @@ public function testSpecificFieldsCanBeSkipped()
     }
+
+
+    #[DataProviderExternal(SearchContextTest::class, 'provideQueryWithinRangeFilter')]
+    public function testQueryWithinRangeFilter(array $params, array $expectedFixtureNames): void
```
The same config should give the same results with both contexts, so makes sense to ensure if the scenarios are updated, both sets of tests run through the updated scenarios.

GuySartorelli commented on tests/php/ORM/DataObjectTest.php:
```diff
@@ -1419,6 +1430,241 @@ public function testSearchableFields()
+    }
+
+    #[DataProvider('provideScaffoldSearchFields')]
+    public function testScaffoldSearchFields(array $config, string $generalSearchFieldName, array $expected): void
```
We had _no_ test coverage for `DataObject::scaffoldSearchFields()`! So I've done reasonably comprehensive test scenarios here.

GuySartorelli commented on tests/php/Forms/GridField/GridFieldFilterHeaderTest/ModelWithBadSearchableFields.php:
```diff
@@ -0,0 +1,34 @@
+    // Explicitly empty
+    private static $searchable_fields = [];
+
+    public function searchableFields()
+    {
+        // Explicitly only include this custom field
+        return [
+            'WhatIsThis' => [
+                'field' => TextField::class,
+                'title' => $this->fieldLabel('WhatIsThis'),
+            ],
+        ];
+    }
```
Needed because you can't have a non-db field in `searchable_fields` but you _can_ have one in this method directly. See https://github.com/silverstripe/silverstripe-framework/pull/11602#discussion_r1951777114

GuySartorelli commented on tests/php/Forms/GridField/GridFieldFilterHeaderTest.php:
```diff
@@ -216,15 +218,16 @@ public function testCanFilterAnyColumns()
         Config::modify()->set(Team::class, 'searchable_fields', ['Name']);
         $this->assertTrue($filterHeader->canFilterAnyColumns($gridField));
 
-        // test that you can filterBy if searchable_fields even if it is not a legit field
-        // this is because we're making a blind assumption it will be filterable later in a SearchContext
-        Config::modify()->set(Team::class, 'searchable_fields', ['WhatIsThis']);
-        $this->assertTrue($filterHeader->canFilterAnyColumns($gridField));
```
See https://github.com/silverstripe/silverstripe-framework/pull/11602#discussion_r1951777114 and https://github.com/silverstripe/silverstripe-framework/pull/11602#discussion_r1951768927

GuySartorelli commented on src/ORM/Filters/WithinRangeFilter.php:
Need these getters for the `match_any` in search context.

GuySartorelli commented on src/ORM/DataObject.php:
```diff
@@ -3896,28 +3933,40 @@ public function searchableFields()
         $rewrite = [];
         foreach ($fields as $name => $specOrName) {
             $identifier = (is_int($name)) ? $specOrName : $name;
+            $relObject = isset($specOrName['match_any']) ? null : $this->relObject($identifier);
```
This was only being called if the spec is an array - but that means if you're trying to search against a field that isn't valid, you only get an exception if you're using complex config! Otherwise you have to filter to get the exception, which isn't very friendly.

Note that it _is_ possible for this to return `null`

emteknetnz reviewed (changes requested):
I followed the example in the developer-docs PR and end up with this somewhat suboptimal UX - the validation errors showed on blur after making no changes to the fields, just clicking between them

I had a "MyDataObject" in a "MyModelAdmin", with things auto-scaffolded

![image](https://github.com/user-attachments/assets/6f82fb6d-d58c-48ed-8f72-ebdb36e785fc)

I only had this PR installed - am I missing something?

GuySartorelli commented:
Was the front-end errors on blur the only thing you encountered? I suspect you'll find that exact same thing happens with that field type when any other filter is used even without this PR (i.e. it's a pre-existing problem). I haven't touched JS for this PR.

It doesn't prevent you submitting the form with or without values in those fields.

emteknetnz commented:
OK, looks like it's CurrencyField that has the onblur issue, also does it with other filters e.g. GreaterThanFilter, there may be others where their default value is considered invalid. Could you please raise a bug issue for this? You probably have more context than me about why it's breaking like it is

I also tried the WithinRangeFilter on an Int field and that mostly worked, there were no onblur validation issues, and the gridfield did filter, however after filtering and opening up the filter again, then values I had put in were no longer there.

GuySartorelli commented:
> Could you please raise a bug issue for this?

Done (https://github.com/silverstripe/silverstripe-admin/issues/1899) - though I haven't looked at all into a) what causes that or b) what the desired UX should be.

> I also tried the WithinRangeFilter on an Int field and that mostly worked, there were no onblur validation issues, and the gridfield did filter, however after filtering and opening up the filter again, then values I had put in were no longer there.

Ahh yup I missed a fix that was originally included in my larger set of PRs. Raised https://github.com/silverstripe/silverstripe-framework/pull/11609 since this is a bug that exists in CMS 5. Will need to be merged, then merged up, and I'll rebase on top.

GuySartorelli commented:
CI failures are being fixed in https://github.com/silverstripe/silverstripe-framework/pull/11615 and aren't related to this PR

emteknetnz reviewed (changes requested):
Tested locally, work good with Date and Int data types

With Currency, which I is what is listed in as an example in the docs, seems very broken e.g. validates default values as invalid, clear button doesn't set to a valid value, the actual filtering doesn't seem to work.  I'm not sure how much of that is an existing issue, or how much of it relates to this PR. Could you have a look and let me know if it should be fixed here or in another issue

GuySartorelli commented:
> the actual filtering doesn't seem to work

Can you please be more specific about that? That sounds like something I need to fix in this PR but I can't without details about what you're doing, what you're expecting, and what's actually happening.

emteknetnz commented:
OK it's specific to this PR with the WithinRangeFilter, not prior to this PR with the GreatThanFilter

Filtering a currency value between two currency values just doesn't work i.e. no results returned

```php
<?php

use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\Filters\WithinRangeFilter;

class MyDataObject extends DataObject
{
    private static $table_name = 'MyDataObject';

    private static $db = [
        'Title' => 'Varchar',
        'MyCurrency' => 'Currency',
    ];

    private static array $searchable_fields = [
        'MyCurrency' => [
            'filter' => WithinRangeFilter::class,
        ],
    ];
}
```

And create ModelAdmin to manage the model

GuySartorelli commented:
Seems to work fine for me?

<details>
<summary>Expand to see the exact PHP I used for the model and admin</summary>

```php
<?php

use SilverStripe\Admin\ModelAdmin;
use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\Filters\WithinRangeFilter;

class MyDataObject extends DataObject
{
    private static $table_name = 'MyDataObject';

    private static $db = [
        'Title' => 'Varchar',
        'MyCurrency' => 'Currency',
    ];

    private static array $summary_fields = [
        'Title',
        'MyCurrency',
    ];

    private static array $searchable_fields = [
        'MyCurrency' => [
            'filter' => WithinRangeFilter::class,
        ],
    ];
}

class MyModelAdmin extends ModelAdmin
{
    private static string $url_segment = 'my-admin';

    private static string $menu_title = 'my admin';

    private static array $managed_models = [
        MyDataObject::class,
    ];
}
```

</details>


https://github.com/user-attachments/assets/d816f500-fd61-4f12-94c1-a7c67adcd5aa


What sort of values are you trying to add? Are you adding the `$` sign before the values, or something like that?

emteknetnz commented:
Ah right, it'll work if I remove the leading $ from the filter values.  This is very confusing as the filter values start off with leading $

GuySartorelli committed:
FIX Pass filter fields through the form before using them.

Some form fields such as CurrencyField manipulate the value to give more
pure values, e.g. removing a currency symbol before the actual value.

GuySartorelli commented:
Turns out the values weren't being passed through the form first, which meant form fields like `CurrencyField` didn't get a chance to update the value - so the DB query was trying to check if `$1.23` matched `1.23` and uhhhh no it doesn't lol.

Added a new commit to pass values through the form first, with a fallback to the original filter value if somehow there's no form field (which shouldn't ever happen but would have been the previous behaviour anyway)

emteknetnz commented on src/Forms/GridField/GridFieldFilterHeader.php:
```diff
@@ -116,9 +116,13 @@ public function handleAction(GridField $gridField, $actionName, $arguments, $dat
+                $this->removeSearchPrefixFromFields($form->Fields());
+                $form->loadDataFrom($filterValues);
+                foreach ($filterValues as $fieldName => $rawFilterValue) {
+                    $state->Columns->$fieldName = $form->Fields()->dataFieldByName($fieldName)?->dataValue() ?? $rawFilterValue;
```
Is search restricted to only database values, or can non-database values also be used?  Is non-database is allowed then dataFieldByName() should be fieldByName()?

Also dataValue() is intended to be value for insertion into the database, getValue() is the internal value of a field, which may be more appropriate?

GuySartorelli replied:
> Is search restricted to only database values, or can non-database values also be used?

For `DataObject` records using the default `SearchContext` only database values can be used - but someone who knows what they're doing could use a custom `SearchContext` - and anyone can use the `BasicSearchContext` to filter on non-db fields.

> Is non-database is allowed then dataFieldByName() should be fieldByName()?

Those methods only change what type of form field is used - e.g. `LiteralField` will be returned from `fieldByName()` but not `dataFieldByName()`, while `TextField` will be the other way around. We _have_ to use `dataFieldByName()` to get fields that can have data - it's nothing to do with the database.

> Also dataValue() is intended to be value for insertion into the database, getValue() is the internal value of a field, which may be more appropriate?

Looking at `CurrencyField`, we _have_ to use `dataValue()` on at least that form field to get a value that is suitable for use here.
Given these values are directly being used in a database query, probably `dataValue()` is the appropriate method since they'll be compared with the values which got stored in the database which also came from `dataValue()`.

emteknetnz reviewed (changes requested):
It's better now, though if I leave the filter values unchanged i.e. $0.00 and $0.00, then it will filter out values.  When the values are left in their default state they should do any filtering

Have talked offline, best solution is probably to swap out CurrencyField for TextField in the method that scaffolds search fields. Have a look at other fields that should probably also have this to prevent formatting or validation that prevents partial searches e.g. MoneyField, EmailField, UrlField, IpField, etc

GuySartorelli committed:
ENH Update scaffolded search fields to be more sensible

GuySartorelli commented:
> swap out CurrencyField for TextField 

I've gone for `NumericField` since the value should never be e.g. "seven dollars" but instead `7.00`

Updated `DBEmail` and `DBUrl` to use `TextField`. `DBIp` already uses that field.

I've left enums alone - they have concrete values, so a dropdown makes sense for them in general (and withinrange probably doesn't make sense for them so they're not really in scope anyway)

To use `DBMoney` you'd have to do something like this to use it (at which point you're not actually dealing with `MoneyField` directly anyway):

```php
<?php

use SilverStripe\ORM\DataObject;
use SilverStripe\ORM\Filters\WithinRangeFilter;

class MyDataObject extends DataObject
{
    private static $table_name = 'MyDataObject';

    private static $db = [
        'Title' => 'Varchar',
        'MyMoney' => 'Money',
    ];

    private static array $summary_fields = [
        'Title',
        'MyMoney',
    ];

    private static array $searchable_fields = [
        'MyMoney.Currency',
        'MyMoney.Amount' => [
            'filter' => WithinRangeFilter::class,
        ],
    ];
}
```
And that only works for building the form - the default search context can't handle that syntax because it expects `.` syntax to be used with relations, not composite fields. So tl;dr handling `DBMoney` isn't in scope for this and has its own pre-existing challenges in relation to filtering that go waaay beyond the scope of this card which is dealing with within range filter.
